### DISCUSSION TITLE
Improved way to get device id.

### DISCUSSION THREAD

**Discussion by GvozdevLeonid** [dashingsoft/pyarmor#1925] [Ideas]
There is a project [plyer](https://github.com/kivy/plyer), which supports getting device id regardless of the platform. It works well and allows you to get the device id and not the hard drive. I checked it on a MacBook and Android - it works great. I work on Android and unfortunately I can't create an external key with a binding to the device.

I also wanted to know if it is possible to keep two runtime files in one project, where one will be perpetual and the second will have an external key file? Will it be possible to replace the runtime with an external key with one that does not have a license? I would like to wrap the main code with a runtime and external key file, but the code that checks for id wrap with perpetual runtime

`pip install plyer`

```
from plyer import uniqueid
print(devicename.device_name)
```

**Comment by jondy** [comment]
Please refer to
https://pyarmor.readthedocs.io/en/latest/tutorial/customization.html#using-hook-to-bind-script-to-docker-id
https://pyarmor.readthedocs.io/en/latest/tutorial/customization.html#using-hook-to-check-network-time-by-other-service
It shows how to check extra device information

**Comment by GvozdevLeonid** [comment]
Oh, I didn't think of that usage. Thanks.

I noticed one unpleasant nuance - the code from bootstrap is included in the license file almost unchanged and it was not difficult for me to understand what exactly is being done. But I found a solution - I encoded almost all the code into a binary string and shifted the bytes by about 55. Now I have an incomprehensible set of bits

**Comment by jondy** [comment]
Try to enable this configuration item
```
pyarmor cfg obf_key_mode 1
```
It could obfuscate the runtime key data also.

**Comment by GvozdevLeonid** [comment]
thanks
